tools/migration/migrator_base.py

```diff
@@ -0,0 +1,242 @@
+from teamscale_client import TeamscaleClient
+from teamscale_client.data import ServiceError
+from requests.exceptions import ConnectionError
+
+
+def get_arguments():
+    """ Parses the arguments for the migration tool. """
+    parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument("config", help="The path to the config file. Needs to be in a specific format, "
+                                       "see config.template.")
+    parser.add_argument("--debug", action="store_true", help="The debug option which enables debug log.")
+    parser.add_argument("--dry-run", action="store_true", help="Dry-run for the migration. See what would happen "
+                                                               "without any consequences")
+    parser.add_argument("--step-by-step", action="store_true", help="Pauses between each migration. Can be used to "
+                                                                    "check for potential error without tainting "
+                                                                    "the complete data.")
+    args = parser.parse_args()
+    return load_config_json(args.config), args.debug, args.dry_run, args.step_by_step
+
+
+def load_config_json(path):
+    """ Loads the config data as a JSON and returns it. """
+    logger = MigratorBase.logger
+    config_file = Path(path)
+    if config_file.exists():
+        try:
+            return json.load(config_file.open())
+        except json.JSONDecodeError:
+            logger.exception("Config file '%s' is malformed" % path, exc_info=True)
+    else:
+        logger.exception("Config file '%s' does not exist" % path)
+    exit(1)
+
+
+def create_logger(name="migrator"):
+    """ Creates a logger """
+    logger = logging.getLogger(name)
+    logger.setLevel(logging.INFO)
+    handler = logging.StreamHandler()
+    formatter = logging.Formatter("%(levelname)-8s %(message)s")
+    handler.setFormatter(formatter)
+    logger.addHandler(handler)
+    return logger
+
+
+class MigratorBase(ABC):
+    """ Base class for migrating data from one instance to another via REST calls. """
+    logger = create_logger()
+
+    def __init__(self, config_data, debug=False, dry_run=False, step_by_step=False):
+        self.debug = debug
+        if self.debug:
+            self.logger.setLevel(logging.DEBUG)
+        else:
+            self.logger.setLevel(logging.INFO)
+        self.dry_run = dry_run
+        self.step_by_step = step_by_step
+        self.old, self.new = self.create_clients(config_data)
+        self.versions_match = self.check_versions()
+        self.check_projects()
+        self.migrated = 0
+        self.cache = {}
+
+        if self.debug:
+            self.logger.debug("Debug Mode ON")
+
+    def check_projects(self):
+        """ Check if the two project actually do exist on the given servers. """
+        try:
+            check_url = "{0.url}/create-project/{0.project}"
```

Comment:
I think the correct call for this is /projects/{id}, see https://git.cqse.eu/cqse/teamscale/blob/master/lib/Cqse.Teamscale.Client/src/Service/Implementation/ProjectService.cs